private acme article (#11)
Kapiche/kapiche.github.io: _posts/2022-04-24-using-private-acme-server-with-caddy-and-private-certs-in-Python.md

---
layout: post
title: "Using Private ACME server with Caddy and private certs in Python"

> description: "How to setup a private acme server to use with caddy and then setup Python to trust the private certificates"

Reviewer: ```suggestion
description: "How to setup a private acme server to use with Caddy and then set up Python to trust the private certificates"
```

category: "Engineering"
author: cam_parry
tags: [Caddy, ACME, Security, Private Certificates, Zero Trust]
share: true
---

Problem:

> In order to request or have automatic lets encrypt with Caddy you can either use HTTP-01 challenge, DNS challeng type or TLS ALPN challenge type. In order to use the HTTP or TLS challenge types you need to have your webserver publicly exposed on port 80 or 443. This worked for awhile but we saw a lot of bots and hack attempts, even though these servers were behind DDoS protection and bot protection services. So we needed something better, something that would take us forward in our zero trust journey. Zero trust network access, as we aren't really validating until we get to our webserver what or who is making the request, so we could do this better. We tried using DNS challenge type but the it required a service account key with pretty much full write access to a DNS zone, which kind of moves the threat vector instead of fixing it.

Reviewer: ```suggestion
In order to request or have automatic Let’s Encrypt with Caddy you can either use HTTP-01 challenge, DNS challenge type or TLS ALPN challenge type. In order to use the HTTP or TLS challenge types you need to have your webserver publicly exposed on port 80 or 443. This worked for a while but we saw many bots and hack attempts, even though these servers were behind DDoS protection and bot protection services. We needed something better, something that would take us forward in our zero trust journey. Zero trust network access, as we aren't really validating until we get to our webserver what or who is making the request, so we could do this better. We tried using the DNS challenge type but it required a service account key with pretty much full write access to a DNS zone, which moves the threat vector instead of fixing it.
```

Journey:
We setup the Google CA and ACME server fairly easily following the instructions https://smallstep.com/docs/registration-authorities/acme-for-cas. The Smallstep ACME RA being available in the Google Cloud marketplace made things really easy.

Now we want to setup firewall rules to only allow our webservers to access the new private ACME RA, so only we can use it. We also want firewall rules to lockdown our webservers our only expose them to the private ACME RA.

We then configure Caddy to use the new private ACME server, instructions given here https://smallstep.com/docs/tutorials/acme-protocol-acme-clients#caddy-v2 or here is our config:

```
    tls {
        issuer acme {
             dir <acme ra server>
             disable_http_challenge
             trusted_roots ./certs/root_ca.crt
        }
    }
```

> We disable the http challenge, so we have only the secure tls alpn challenge.

Reviewer: ```suggestion
We disable the HTTP challenge, so we have only the secure TLS alpn challenge.
```

We can now use this Private ACME RA server for all the other supported acme clients https://smallstep.com/docs/tutorials/acme-protocol-acme-clients including Kubernetes cert-manager.

We have our private certificates setup and working with Caddy, but the microservices that talk to the webservers need the private certificate in their certificate store in order to send verified TLS requests.

> We found a few old blog about using private certificates with python but they didn't quite work, either they were older blog articles or we must not have been doing something correctly. There is one issue raised in python requests library which made things a little more nuanced https://github.com/psf/requests/issues/2966. Basically you have to workaround the use of certifi and how it handles certificates. 

Reviewer: ```suggestion
We found a few old blog about using private certificates with Python but they didn't quite work, either they were older blog articles or we must not have been doing something correctly. There is one issue raised in Python requests library which made things a little more nuanced https://github.com/psf/requests/issues/2966. Basically you have to work around the use of certifi and how it handles certificates. 
```
